### PR TITLE
BP-1253: review ensure_process_kill() - ported from movai_classes (BP-1241)

### PR DESCRIPTION
**Pull request by andreparames**
Ported from b46b0dbc (https://github.com/MOV-AI/movai_classes/pull/189/)

TIcket: [BP-1253](https://movai.atlassian.net/browse/BP-1253)

See also https://github.com/MOV-AI/movai-core-shared/pull/167

# content
- review termination of nodes by sigterm, then sigint, then sigkill, 2 functions modified:
  - ensure_process_kill(): full logic was reviewed and explained in docstrings and comments
  - terminate_process(): small adaptation to the new prototype of ensure_process_kill()

# test

```
import signal
import time
import multiprocessing

fvar = Var('Flow')
flag_var_name = "test_sigterm_running"

# Set the flag to 1
fvar.set(flag_var_name, 1)


def signal_handler(signum, frame):
    logger.info("SIGTERM received, cleaning up...")
    # stop the subprocess
    fvar = Var('Flow')
    fvar.set("test_sigterm_running", 0)


def print_and_sleep(sleep_time=60):
    """ Print a message and sleep for a given amount of time """
    import time
    fvar = Var('Flow')

    running_flag = fvar.get("test_sigterm_running")

    logger.info(f"Sleeping for {sleep_time} seconds")
    while running_flag > 0 and sleep_time > 0:
        logger.info(f"Sleeping for 1s, {sleep_time} remaining")
        sleep_time -= 1
        time.sleep(1)
        running_flag = fvar.get("test_sigterm_running")

    logger.info("Exiting subprocess")

# Set up signal handling
signal.signal(signal.SIGTERM, signal_handler)

# create a subprocess that will hang indefinitely
p = multiprocessing.Process(target=print_and_sleep, args=(60,))
p.start()

time.sleep(3)

# Exit
gd.oport['done'].send()

```

The callback below when used in a transitionIn of a Node, would prevent the flow to correctly transition after its Node. 
The fix proposed was tested with this flow:
image

The PR also includes unit tests to confirm the behavior of the changes.
---
- [x] Make sure you are opening from a **topic/feature/bugfix branch**
- [x] Ensure that the PR title represents the desired changes
- [x] Ensure that the PR description detail the desired changes
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

[^note]:
    Put an `x` into the [ ] to show you have filled the information.
    The template comes from https://github.com/MOV-AI/.github/blob/master/.github/pull_request_template.md
    You can override it by creating .github/pull_request_template.md  in your own repository

[BP-1253]: https://movai.atlassian.net/browse/BP-1253?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: flow_initiator/spawner/elements/process_element.py [hunk 1]
```diff
@@ -76,7 +82,12 @@ def __del__(self):
...
-        if self.proc.returncode is None:
+        if not self.proc:
+            self._logger.info(
+                "ProcessElement deleted before process was started (node: %s)",
+                self.node_name,
+            )
+        elif self.proc.returncode is None:
             self.proc.kill()
 
     async def wait(self):
```
Comment (AlexFernandes-MOVAI) [inline flow_initiator/spawner/elements/process_element.py:93]: Function below should be reviewed:
```
async def kill(self):
        """
        A kill function for stopping
        Returns: None
        """
        self._logger.info(
            "Spawner: Trying to terminate process({})".format(self.proc.pid)
        )
        await self.ensure_process_kill()
        
```

**Review by AlexFernandes-MOVAI — COMMENTED** [review]
Difference of behavior with 2.3.1:
- sigterm signal is not catched by the node callback

Can be due to the python version ? python modules version ? or `proc.terminate()` instead of `send_signal(signal.SIGTERM)`

**Comment by andreparames** [comment]
> Difference of behavior with 2.3.1:
> 
>     * sigterm signal is not catched by the node callback
> 
> 
> Can be due to the python version ? python modules version ? or `proc.terminate()` instead of `send_signal(signal.SIGTERM)`

I think my port missed the part where the actual SIGTERM was sent. It used to be sent in terminate_process(), but that doesn't exist in the new code. Instead, I made ensure_process_kill() send SIGTERM after the SIGINT (and before the SIGKILL).

**Review by AlexFernandes-MOVAI — CHANGES_REQUESTED** [review]
Sequence is still incorrect, SIGINT is sent before SIGTERM

```
[INFO][2024-10-28 10:58:04][node][initial_print][118]: --- Starting GD_Node with Name "sleep_task" and Template "sleep_task"
[INFO][2024-10-28 10:58:04][zmq_client][_reset][61]: ZMQ connecting DEALER to: ipc:///opt/mov.ai/comm/SpawnerServer-basic_simulator-movai.sock
[INFO][2024-10-28 10:58:04][sigterm_sleep][print_and_sleep][29]: [user_log:True|node:sleep_task|callback:sigterm_sleep] Sleeping for 60 seconds
[INFO][2024-10-28 10:58:04][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 60 remaining
[INFO][2024-10-28 10:58:05][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 59 remaining
[INFO][2024-10-28 10:58:06][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 58 remaining
[INFO][2024-10-28 10:58:07][spawner][process_transition][518]: TRANS from node: sleep_task   port: end
[INFO][2024-10-28 10:58:07][flowmonitor][get_commands][149]: load flow __UNVERSIONED__
[INFO][2024-10-28 10:58:07][process_element][kill][105]: Spawner: Trying to terminate process(1623)
[INFO][2024-10-28 10:58:07][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 57 remaining
[INFO][2024-10-28 10:58:08][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 56 remaining
[WARNING][2024-10-28 10:58:09][process_element][ensure_process_kill][179]: Sending SIGINT to node sleep_task (1623)
[INFO][2024-10-28 10:58:09][node][stop][89]: Shutting down GD Node: sleep_task
[INFO][2024-10-28 10:58:09][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 55 remaining
[WARNING][2024-10-28 10:58:09][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
[INFO][2024-10-28 10:58:09][topics][_TopicManager.__init__][1132]: topicmanager initialized
[INFO][2024-10-28 10:58:10][node][initial_print][118]: --- Starting GD_Node with Name "wait_state" and Template "wait_state"
[INFO][2024-10-28 10:58:10][zmq_client][_reset][61]: ZMQ connecting DEALER to: ipc:///opt/mov.ai/comm/SpawnerServer-basic_simulator-movai.sock
[WARNING][2024-10-28 10:58:10][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
[INFO][2024-10-28 10:58:10][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 54 remaining
[WARNING][2024-10-28 10:58:10][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
[WARNING][2024-10-28 10:58:11][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
[INFO][2024-10-28 10:58:11][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 53 remaining
[WARNING][2024-10-28 10:58:11][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
[ERROR][2024-10-28 10:58:12][process_element][ensure_process_kill][191]: Sending SIGKILL to node sleep_task (1623)
[INFO][2024-10-28 10:58:12][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 52 remaining
[WARNING][2024-10-28 10:58:12][spawner][clean_element][262]: Node: sleep_task is still running
[ERROR][2024-10-28 10:58:12][spawner][clean_element][281]: **  Node sleep_task (machine: HOST, PID: 1623) terminated with code -9  **
[INFO][2024-10-28 10:58:13][spawner][process_stop][478]: STOP flow global/Flow/_test/__UNVERSIONED__
[INFO][2024-10-28 10:58:13][process_element][kill][105]: Spawner: Trying to terminate process(1648)
[INFO][2024-10-28 10:58:13][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 51 remaining
[INFO][2024-10-28 10:58:14][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 50 remaining
```

**Comment by AlexFernandes-MOVAI** [comment]
> Sequence is still incorrect, SIGINT is sent before SIGTERM
> 
> ```
> [INFO][2024-10-28 10:58:04][node][initial_print][118]: --- Starting GD_Node with Name "sleep_task" and Template "sleep_task"
> [INFO][2024-10-28 10:58:04][zmq_client][_reset][61]: ZMQ connecting DEALER to: ipc:///opt/mov.ai/comm/SpawnerServer-basic_simulator-movai.sock
> [INFO][2024-10-28 10:58:04][sigterm_sleep][print_and_sleep][29]: [user_log:True|node:sleep_task|callback:sigterm_sleep] Sleeping for 60 seconds
> [INFO][2024-10-28 10:58:04][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 60 remaining
> [INFO][2024-10-28 10:58:05][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 59 remaining
> [INFO][2024-10-28 10:58:06][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 58 remaining
> [INFO][2024-10-28 10:58:07][spawner][process_transition][518]: TRANS from node: sleep_task   port: end
> [INFO][2024-10-28 10:58:07][flowmonitor][get_commands][149]: load flow __UNVERSIONED__
> [INFO][2024-10-28 10:58:07][process_element][kill][105]: Spawner: Trying to terminate process(1623)
> [INFO][2024-10-28 10:58:07][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 57 remaining
> [INFO][2024-10-28 10:58:08][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 56 remaining
> [WARNING][2024-10-28 10:58:09][process_element][ensure_process_kill][179]: Sending SIGINT to node sleep_task (1623)
> [INFO][2024-10-28 10:58:09][node][stop][89]: Shutting down GD Node: sleep_task
> [INFO][2024-10-28 10:58:09][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 55 remaining
> [WARNING][2024-10-28 10:58:09][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
> [INFO][2024-10-28 10:58:09][topics][_TopicManager.__init__][1132]: topicmanager initialized
> [INFO][2024-10-28 10:58:10][node][initial_print][118]: --- Starting GD_Node with Name "wait_state" and Template "wait_state"
> [INFO][2024-10-28 10:58:10][zmq_client][_reset][61]: ZMQ connecting DEALER to: ipc:///opt/mov.ai/comm/SpawnerServer-basic_simulator-movai.sock
> [WARNING][2024-10-28 10:58:10][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
> [INFO][2024-10-28 10:58:10][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 54 remaining
> [WARNING][2024-10-28 10:58:10][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
> [WARNING][2024-10-28 10:58:11][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
> [INFO][2024-10-28 10:58:11][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 53 remaining
> [WARNING][2024-10-28 10:58:11][process_element][ensure_process_kill][179]: Sending SIGTERM to node sleep_task (1623)
> [ERROR][2024-10-28 10:58:12][process_element][ensure_process_kill][191]: Sending SIGKILL to node sleep_task (1623)
> [INFO][2024-10-28 10:58:12][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 52 remaining
> [WARNING][2024-10-28 10:58:12][spawner][clean_element][262]: Node: sleep_task is still running
> [ERROR][2024-10-28 10:58:12][spawner][clean_element][281]: **  Node sleep_task (machine: HOST, PID: 1623) terminated with code -9  **
> [INFO][2024-10-28 10:58:13][spawner][process_stop][478]: STOP flow global/Flow/_test/__UNVERSIONED__
> [INFO][2024-10-28 10:58:13][process_element][kill][105]: Spawner: Trying to terminate process(1648)
> [INFO][2024-10-28 10:58:13][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 51 remaining
> [INFO][2024-10-28 10:58:14][sigterm_sleep][print_and_sleep][31]: [user_log:True|node:sleep_task|callback:sigterm_sleep] (1639)Sleeping for 1s, 50 remaining
> ```

@andreparames  did you investigate the fact that the children processes do not receive the sigterm in comparison to 2.3.1 ?

**Pull request merged**